File: SampleCode/Android/app/src/main/java/com/netease/yunxin/kit/meeting/sampleapp/data/ServerConfig.kt
/*
 * Copyright (c) 2022 NetEase, Inc. All rights reserved.
 * Use of this source code is governed by a MIT license that can be
 * found in the LICENSE file.
 */

package com.netease.yunxin.kit.meeting.sampleapp.data
import com.netease.yunxin.kit.meeting.sampleapp.utils.SPUtils

data class ServerConfig(
    val name: String,
    val appKey: String,
    val serverUrl: String,
    val isDebugMode: Boolean = false,
    val useAssetServerConfig: Boolean = false
)

object ServerConfigs {

    private val SERVER_CONFIGS: Map<String, ServerConfig> = listOf(
        ServerConfig(
            name = "online",
            appKey = "91d597b20132e6fa131615aa2d229388",
            serverUrl = ""
        ),
        ServerConfig(
            name = "test",
            appKey = "4649991c6ab7cc5a4309ccf25d8793e5",
            serverUrl = "https://roomkit-dev.netease.im/",
            isDebugMode = true
        ),
        ServerConfig(
            name = "private",
            appKey = "ff965d13edd293d22bcf78af1a621bcb",
            serverUrl = "",
            useAssetServerConfig = true
        )
    ).associateBy { it.name }

    fun determineServerConfig(appKey: String): ServerConfig {
        // val serverConfig = ServerConfig(appKey = appKey, appServerUrl = "", sdkServerUrl = "")
        val serverConfig = SERVER_CONFIGS[SPUtils.getInstance().getString("server_type", "test")]!!

        return serverConfig.copy(
            appKey = SPUtils.getInstance().getString("custom_app_key").ifEmpty { serverConfig.appKey },
            serverUrl = SPUtils.getInstance().getString("custom_sdk_server_url").ifEmpty { serverConfig.serverUrl }
        )
    }
}

File: SampleCode/iOS/NEMeetingDemo/ServerConfig.m
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

#import "ServerConfig.h"
#import "LoginInfoManager.h"

NSString *const kAppKey = @"Your_Meeting_App_Key";

NSString * const kServerType = @"serverType";
NSString * const kCustomAppKey = @"customAppKey";
NSString * const kCustomAppServerUrl = @"customAppServerUrl";
NSString * const kCustomSDKServerUrl = @"customSDKServerUrl";


@implementation ServerConfig


+ (NSDictionary<NSString *,ServerConfig *> *)servers {
    static NSDictionary *configs;
    if (configs == nil) {
        configs = @{
            @"online": [[ServerConfig alloc] init: kAppKey sdkServerUrl: @"你的sdkServerUrl"],
        };
    }
    return configs;
}

- (instancetype)init: (NSString*) appkey
        sdkServerUrl: (NSString*) sdkServerUrl
{
    self = [super init];
    if (self) {
        self.appKey = appkey;
        self.sdkServerUrl = sdkServerUrl;
    }
    return self;
}

+ (ServerConfig *)current {
    static ServerConfig * currentConfig;
    if (currentConfig == nil) {
        ServerConfig* candicate = [ServerConfig servers][[ServerConfig serverType]];
        
        NSString* appKey = [ServerConfig ifEmpty:[ServerConfig customAppKey] fallback: candicate.appKey];
        NSString* sdkServerUrl = [ServerConfig ifEmpty:[ServerConfig customSDKServerUrl] fallback:candicate.sdkServerUrl];
        
        currentConfig = [[ServerConfig alloc] init: appKey sdkServerUrl: sdkServerUrl];
        
        NSLog(@"Select server config: appKey: %@, sdkServerUrl: %@", appKey, sdkServerUrl);
    }
    
    return currentConfig;
}

+ (NSString *) ifEmpty: (NSString*) value fallback: (nonnull NSString*) fallback  {
    if (value == nil || value.length == 0) {
        return fallback;
    }
    return value;
}

+ (NSString*)serverType {
    return [ServerConfig ifEmpty:[[NSUserDefaults standardUserDefaults] stringForKey: kServerType] fallback: @"online"];
}

+ (void)setServerType:(NSString*)serverType {
    [[NSUserDefaults standardUserDefaults] setValue:serverType forKey:kServerType];
}

+ (NSString *)customAppKey {
    return [[NSUserDefaults standardUserDefaults] stringForKey: kCustomAppKey];
}

+ (void)setCustomAppKey:(NSString *)customAppKey {
    if (customAppKey && (NSNull *)customAppKey != [NSNull null]) {
        [[NSUserDefaults standardUserDefaults] setValue:customAppKey forKey:kCustomAppKey];
    } else {
        [[NSUserDefaults standardUserDefaults] removeObjectForKey:kCustomAppKey];
    }
}

+ (NSString *)customAppServerUrl {
    return [[NSUserDefaults standardUserDefaults] stringForKey: kCustomAppServerUrl];
}

+ (void)setCustomAppServerUrl:(NSString *)customAppServerUrl {
    if (customAppServerUrl && (NSNull *)customAppServerUrl != [NSNull null]) {
        [[NSUserDefaults standardUserDefaults] setValue:customAppServerUrl forKey:kCustomAppServerUrl];
    } else {
        [[NSUserDefaults standardUserDefaults] removeObjectForKey:kCustomAppServerUrl];
    }
}

+ (NSString *)customSDKServerUrl {
    return [[NSUserDefaults standardUserDefaults] stringForKey: kCustomSDKServerUrl];
}

+ (void)setCustomSDKServerUrl:(NSString *)customSDKServerUrl {
    if (customSDKServerUrl && (NSNull *)customSDKServerUrl != [NSNull null]) {
        [[NSUserDefaults standardUserDefaults] setValue:customSDKServerUrl forKey:kCustomSDKServerUrl];
    } else {
        [[NSUserDefaults standardUserDefaults] removeObjectForKey:kCustomSDKServerUrl];
    }
}

@end

